pdfparser/poppler.pyx

```diff
@@ -575,134 +577,211 @@ cdef class LinesIterator:
 
 cdef class Line:
     cdef:
-        PopplerPage page
-        TextLine *line
-        double x1, y1, x2, y2
-        unicode _text
-        list _bboxes
-        CompactList _fonts
+        PopplerPage  page
+        TextLine    *line
+        unicode      text
+        BBox         bbox
+        list         char_bboxes
+        CompactList  char_fonts
 
     def __cinit__(self, PopplerPage page):
         self.page = page
         self.line = NULL
+        self.text = None
 
     cdef Line wrap(self, TextLine *line):
         self.line = line
-        self._text = u'' # text bytes
-        self.x1 = 0
-        self.y1 = 0
-        self.x2 = 0
-        self.y2 = 0
-        self._bboxes = []
-        self._fonts = CompactList()
-        self._get_text()
         return self
 
-    def _get_text(self):
+    cdef TextWord *getWords(self):
+        if self.line:
+            return self.line.getWords()
+        else:
+            return NULL
+
+    def __iter__(self):
+        return WordsIterator(self.page, self)
+
+    cdef _cache_line_data(self):
         cdef:
-            TextWord *word
-            GooString *_word_text
-            GooString *_font_name
-            double bx1, bx2, by1, by2
-            list words = []
-            int i, word_length
-            BBox last_bbox
-            FontInfo last_font
-            double r, g, b
+            Word word
+            list bboxes
+            list word_texts = []
+            list char_bboxes = []
+            CompactList char_fonts = CompactList()
+            double l = DBL_MAX
+            double t = DBL_MAX
+            double r = DBL_MIN
+            double b = DBL_MIN
+
+        add_word_space = False
+
+        for word in self:
+            # skip empty words (if any)
+            if len(word.text) == 0:
+                continue
+
+            bboxes = word.char_bboxes
+
+            # if the previous word ended with space,
+            # it needs to be added because it is not
+            # a part of the word
+            if add_word_space:
+                word_texts.append(u' ')
+                char_bboxes.append(BBox(char_bboxes[-1].x2,
```

Comment:
What about `char_bboxes[-1]` in the first iteration when `char_bboxes` is empty?

Reply:
Aha, that can happen only if `add_word_space == True`, ie. not in the first iteration.

Reply:
Would be perfect to comment that space left = previous word right, space right = current word left. Possibly extract `prev_word_bbox = char_bboxes[-1]` and `word_bbox = bboxes[0]`.